response_operations_ui/views/collection_exercise.py

```diff
@@ -1026,3 +1050,68 @@ def _delete_seft_collection_instrument(short_name, period):
             success_panel=success_panel,
         )
     )
+
+
+def _add_collection_instrument(short_name, period):
+    form = LinkCollectionInstrumentForm(form=request.form)
+    try:
+        # The eq_id of a collection instrument will ALWAYS be its shortname.
+        short_name_lower = str(short_name).lower()
+        survey_uuid = survey_controllers.get_survey_by_shortname(short_name_lower)["id"]
+        collection_instrument_controllers.get_collection_instruments_by_classifier(ci_type="EQ", survey_id=survey_uuid)
+        if not form.validate():
+            unique_error = list(dict.fromkeys(form.formtype.errors))
+
+            session["error"] = json.dumps(
+                {
+                    "section": "add-ci-error",
+                    "header": unique_error[0],
+                    "message": unique_error[0],
+                }
+            )
+            return get_view_sample_ci(short_name, period)
+
+        collection_instrument_controllers.link_collection_instrument_to_survey(
+            survey_uuid, short_name_lower, form.formtype.data
+        )
+        # Need to get selectors a second time as we just added one and the list from before is outdated.
+        collection_instrument_controllers.get_collection_instruments_by_classifier(ci_type="EQ", survey_id=survey_uuid)
+    except ApiError as err:
+        try:
+            session["error"] = json.dumps(
+                {
+                    "section": "add-ci-error",
+                    "header": "Cannot upload an instrument with an identical set of classifiers",
+                    "message": "Enter a CI that has not already been added",
+                }
+            )
+        except (JSONDecodeError, KeyError):
+            # If the message isn't JSON, or the 'errors' key doesn't exist, we'll render the message anyway as it
+            # might still be helpful.
+            errors = [("", [err.message])]
+            session["error"] = json.dumps(errors[0])
+
+        return get_view_sample_ci(short_name, period)
+
+    form.formtype.data = ""  # Reset the value on successful submission
+    return get_view_sample_ci(short_name, period)
+
+
+# Need to create a list containing all the available CIS to be used on the CI selection page. This iss all the
+# CIs already linked and CIs unlinked
+def _get_collective_cis(ce_details, eq_ci_selectors):
+    complete_ci_dict = []
+    duplicate_list = []
+
+    if "EQ" in ce_details["collection_instruments"]:
```

Comment:
Updated.

Reply:
This has been updated to a get

Reply:
I think this double filter process is a bit unnecessary. So filter_eq_ci_selectors is not used for anything other than populating this now, so we should repurpose it, something like 

```
def filter_eq_ci_selectors(eq_ci_selectors: list[dict], collection_instruments: list[dict]) -> list[dict]:
 """ I removed the doc because of review  """    
all_cis_for_survey =[]
    for ci in eq_ci_selectors:
        ci = {
            "id": ci["id"],
            "form_type": ci["classifiers"]["form_type"],
            "checked": "true" if ci in collection_instruments else "false"
        }
        all_cis_for_survey.append(ci)
        
    return all_cis_for_survey 
```
This will form the exact list you had before in all_cis_for_survey, naturally you will need to adjust the render_template vars